### PR TITLE
Support custom vendor strings.

### LINKED ISSUE #33: failure with musl libc
20:49.44      Running `/var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/ff/release/build/target-lexicon-f3f06bed553e79f7/build-script-build`
20:49.46 [target-lexicon 0.8.1] thread 'main' panicked at 'can't find custom target x86_64-gentoo-linux-musl', /var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/third_party/rust/target-lexicon/build.rs:100:5
20:49.46 [target-lexicon 0.8.1] stack backtrace:
20:49.46 [target-lexicon 0.8.1]    0:     0x563128709e21 - std::sys_common::backtrace::print::h287c8c07a68eede3
20:49.46 [target-lexicon 0.8.1]    1:     0x56312870d6c2 - std::panicking::default_hook::{{closure}}::h05f6680debc806b6
20:49.46 [target-lexicon 0.8.1]    2:     0x56312870d460 - std::panicking::default_hook::h15544f9c95b669ca
20:49.46 [target-lexicon 0.8.1]    3:     0x56312870dcfd - std::panicking::rust_panic_with_hook::h3c1da02a49e9001b
20:49.46 [target-lexicon 0.8.1]    4:     0x56312870d892 - std::panicking::continue_panic_fmt::h1225421fdef1e090
20:49.46 [target-lexicon 0.8.1]    5:     0x56312870d7df - std::panicking::begin_panic_fmt::h6cf338a46e92138b
20:49.46 [target-lexicon 0.8.1]    6:     0x5631286ffe28 - build_script_build::main::h2d1c6843c8761064
20:49.47 [target-lexicon 0.8.1]    7:     0x5631286f9173 - std::rt::lang_start::{{closure}}::h2de0c3e376b0d62e
20:49.47 [target-lexicon 0.8.1]    8:     0x56312870d763 - std::panicking::try::do_call::h303cc4ae6bb6cb3c
20:49.47 [target-lexicon 0.8.1]    9:     0x56312870e4ba - __rust_maybe_catch_panic
20:49.47 [target-lexicon 0.8.1]   10:     0x563128706e54 - std::rt::lang_start_internal::he20f95ac4ef5d8da
20:49.47 [target-lexicon 0.8.1]   11:     0x563128702b12 - main
20:49.47 [target-lexicon 0.8.1]   12:     0x7fa1617c9b55 - <unknown>
20:49.47 error: failed to run custom build command for `target-lexicon v0.8.1`
20:49.48 Caused by:
20:49.48   process didn't exit successfully: `/var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/ff/release/build/target-lexicon-f3f06bed553e79f7/build-script-build` (exit code: 101)
20:49.48 --- stderr
20:49.48 thread 'main' panicked at 'can't find custom target x86_64-gentoo-linux-musl', /var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/third_party/rust/target-lexicon/build.rs:100:5
20:49.48 stack backtrace:
20:49.48    0:     0x563128709e21 - std::sys_common::backtrace::print::h287c8c07a68eede3
20:49.48    1:     0x56312870d6c2 - std::panicking::default_hook::{{closure}}::h05f6680debc806b6
20:49.48    2:     0x56312870d460 - std::panicking::default_hook::h15544f9c95b669ca
20:49.48    3:     0x56312870dcfd - std::panicking::rust_panic_with_hook::h3c1da02a49e9001b
20:49.49    4:     0x56312870d892 - std::panicking::continue_panic_fmt::h1225421fdef1e090
20:49.49    5:     0x56312870d7df - std::panicking::begin_panic_fmt::h6cf338a46e92138b
20:49.49    6:     0x5631286ffe28 - build_script_build::main::h2d1c6843c8761064
20:49.49    7:     0x5631286f9173 - std::rt::lang_start::{{closure}}::h2de0c3e376b0d62e
20:49.49    8:     0x56312870d763 - std::panicking::try::do_call::h303cc4ae6bb6cb3c
20:49.49    9:     0x56312870e4ba - __rust_maybe_catch_panic
20:49.49   10:     0x563128706e54 - std::rt::lang_start_internal::he20f95ac4ef5d8da
20:49.49   11:     0x563128702b12 - main
20:49.49   12:     0x7fa1617c9b55 - <unknown>

### PR DESCRIPTION
**Pull request by sunfishcode**
Add support for custom vendors, as in "x86_64-gentoo-linux-musl".

Fixes #33.

### CODE REVIEW

File: src/targets.rs [hunk 1]
```diff
@@ -738,7 +750,46 @@ impl FromStr for Vendor {
...
+                }
+
+                // Require the first character to be an ascii lowercase.
+                if !custom.chars().nth(0).unwrap().is_ascii_lowercase() {
+                    return Err(());
+                }
+
+                // Restrict the set of characters permitted in a custom vendor.
+                if custom
+                    .find(|c: char| {
```
Comment (bjorn3) [inline src/targets.rs]: .any

Comment (sunfishcode) [reply src/targets.rs]: done

**Comment by anarchpenguin** [comment]
17:04.03 error[E0308]: mismatched types
17:04.03  --> /var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/ff/x86_64-gentoo-linux-musl/release/build/target-lexicon-dfbcc777324844a3/out/host.rs:9:28
17:04.03   |
17:04.04 9 |     vendor: Vendor::Custom("gentoo"),
17:04.04   |                            ^^^^^^^^ expected struct `alloc::boxed::Box`, found reference
17:04.04   |
17:04.04   = note: expected type `alloc::boxed::Box<alloc::string::String>`
17:04.04              found type `&'static str`
17:04.24 [style 0.0.1] cargo:rerun-if-changed=build_gecko.rs
17:04.24 [style 0.0.1] cargo:rerun-if-changed=/var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/ff/layout/style/bindgen.toml
17:04.24 [style 0.0.1] cargo:rerun-if-changed=/var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/layout/style/ServoBindings.toml
17:04.26 error[E0308]: mismatched types
17:04.26   --> /var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/ff/x86_64-gentoo-linux-musl/release/build/target-lexicon-dfbcc777324844a3/out/host.rs:25:24
17:04.26    |
17:04.26 25 |         Vendor::Custom("gentoo")
17:04.26    |                        ^^^^^^^^ expected struct `alloc::boxed::Box`, found reference
17:04.26    |
17:04.26    = note: expected type `alloc::boxed::Box<alloc::string::String>`
17:04.26               found type `&'static str`
17:04.28 error[E0308]: mismatched types
17:04.29   --> /var/tmp/portage/www-client/firefox-71.0_beta6/work/firefox-71.0/ff/x86_64-gentoo-linux-musl/release/build/target-lexicon-dfbcc777324844a3/out/host.rs:55:36
17:04.29    |
17:04.29 55 |             vendor: Vendor::Custom("gentoo"),
17:04.29    |                                    ^^^^^^^^ expected struct `alloc::boxed::Box`, found reference
17:04.29    |
17:04.29    = note: expected type `alloc::boxed::Box<alloc::string::String>`
17:04.29               found type `&'static str`
17:04.29 error: aborting due to 3 previous errors

**Comment by anarchpenguin** [comment]
Dan what branch are you working with, I have tried to test last patch with current build in firefox-71.xb* but it does not apply.

**Comment by bnjbvr** [comment]
@anarchpenguin We're manually importing Cranelift and all its dependencies in Mozilla central manually, at the moment, so both code bases might be drifting apart.

If your intent is just to compile Firefox but you don't mind missing Cranelift support, you can use the configure flag `--disable-cranelift` to remove using it locally. For what it's worth, it isn't supposed to be enabled on non-nightly versions, I see you're compiling a beta version, so if you didn't explicitly opt in with `--enable-cranelift`, there might be something weird happening on Gecko's build system. If that's so, can you open a bug on Mozilla's Bugzilla, please?

**Comment by anarchpenguin** [comment]
https://gist.github.com/anarchpenguin/b169c55b00a8fe6c3c5c8c88a8d10ee3

Is the patch I am using with firefox-71.0b*, it works as expected. This is your work just backported onto the 0.8.1 branch thanks for all the support I believe it would be save to merge to master.

File: src/targets.rs [hunk 2]
```diff
@@ -738,7 +784,44 @@ impl FromStr for Vendor {
...
+                // Require the first character to be an ascii lowercase.
+                if !custom.chars().nth(0).unwrap().is_ascii_lowercase() {
+                    return Err(());
+                }
+
+                // Restrict the set of characters permitted in a custom vendor.
+                fn is_prohibited_char(c: char) -> bool {
+                    !(c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_' || c == '.')
+                }
+                if custom.chars().any(is_prohibited_char) {
```
Comment (bnjbvr) [inline src/targets.rs]: nit: `is_prohibited_char` could just be a closure, to avoid the local function definition (used only once)?

Comment (sunfishcode) [reply src/targets.rs]: Done.

**Pull request merged**